File: src/backups/store.js
'use strict';
/**
 * Where backups live on disk, and which ones stop living there.
 *
 * ── A pair is the unit ─────────────────────────────────────────────────────
 *
 *     /data/config-backups/<router-slug>/
 *         2026-08-19T203521.rsc.gz      gzipped export — diffable, no secrets
 *         2026-08-19T203521.backup      aes-sha256 binary — restorable
 *
 * The two files share a stem and are always created, removed and counted
 * together. A `.rsc` without its `.backup` is a diff you cannot act on; a
 * `.backup` without its `.rsc` is a restore point you cannot inspect. Neither
 * half is useful alone, so neither is kept alone.
 *
 * `/export` masks private keys (verified against a live AX3: only WireGuard
 * *public* keys appear), so the `.rsc` carries no secrets and is stored
 * plainly. It still describes the whole network, which is why reading one
 * needs page write rather than page read.
 *
 * ── The directory name, and why the database does not trust it ─────────────
 *
 * The directory is named from the router's label because a human browsing
 * `/data` should be able to tell whose backups these are. Labels change, so
 * the *database* records the directory that was used; nothing resolves an old
 * backup by re-deriving a slug from the current label. Renaming a router
 * starts a new directory and leaves the old one findable.
 */

const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

/**
 * Resolved per call, not frozen at require time.
 *
 * A constant here meant the test suite wrote real backup pairs into the real
 * /data/config-backups — it left a `test-router` directory on a production
 * instance, which is how this was found. Reading the environment when the path
 * is actually needed lets a test point DATA_DIR somewhere disposable.
 */
function baseDir() {
  return path.join(process.env.DATA_DIR || '/data', 'config-backups');
}

/** Anything outside this is replaced, so a label can never escape the base directory. */
const _UNSAFE = /[^a-z0-9]+/g;

/**
 * A directory name from a router label.
 *
 * Deliberately lossy: lowercase, dashes, nothing else. A label of only
 * punctuation would slug to nothing, so it falls back to a fixed name rather
 * than producing '' and writing into the base directory itself.
 */
function slugFor(label) {
  const slug = String(label == null ? '' : label)
    .toLowerCase()
    .replace(_UNSAFE, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 60);
  return slug || 'router';
}

/** Absolute path of a router's backup directory. */
function dirFor(slug) {
  return path.join(baseDir(), slug);
}

/**
 * The filename stem for a moment in time, in UTC.
 *
 * UTC because a local-time stem repeats itself for an hour every autumn, and
 * two backups that sort as equal are two backups that can overwrite each
 * other. Seconds are included for the same reason at a smaller scale: a
 * scheduled run and a manual one can land in the same minute.
 */
function stemFor(ts) {
  const d = new Date(ts);
  const p = (n, w) => String(n).padStart(w || 2, '0');
  return d.getUTCFullYear() + '-' + p(d.getUTCMonth() + 1) + '-' + p(d.getUTCDate()) +
         'T' + p(d.getUTCHours()) + p(d.getUTCMinutes()) + p(d.getUTCSeconds());
}

function rscPath(dir, stem)    { return path.join(dir, stem + '.rsc.gz'); }
function backupPath(dir, stem) { return path.join(dir, stem + '.backup'); }

function ensureDir(dir) {
  fs.mkdirSync(dir, { recursive: true });
}

/**
 * Write both halves of a pair.
 *
 * The binary lands first. If the process dies between the two writes, what
 * survives is a `.backup` with no `.rsc` — and `listPairs()` ignores it,
 * because a half-written pair is not a backup. The other order would leave a
 * diffable export that claims a restore point exists.
 */
function writePair(dir, stem, rscText, backupBuf) {
  ensureDir(dir);
  const gz = zlib.gzipSync(Buffer.from(rscText, 'utf8'));
  fs.writeFileSync(backupPath(dir, stem), backupBuf);
  fs.writeFileSync(rscPath(dir, stem), gz);
  return {
    rscPath: rscPath(dir, stem),
    backupPath: backupPath(dir, stem),
    rscBytes: gz.length,
    backupBytes: backupBuf.length,
  };
}

/** The stored export, gunzipped. */
function readRsc(dir, stem) {
  return zlib.gunzipSync(fs.readFileSync(rscPath(dir, stem))).toString('utf8');
}

/** The stored binary, as it will be pushed back to the router. */
function readBackup(dir, stem) {
  return fs.readFileSync(backupPath(dir, stem));
}

function hasPair(dir, stem) {
  return fs.existsSync(rscPath(dir, stem)) && fs.existsSync(backupPath(dir, stem));
}

/**
 * Remove a pair. Missing halves are not an error — pruning has to be
 * idempotent, because it runs after a crash as readily as after a success.
 */
function removePair(dir, stem) {
  let removed = 0;
  for (const p of [rscPath(dir, stem), backupPath(dir, stem)]) {
    try { fs.unlinkSync(p); removed++; }
    catch (e) { if (e.code !== 'ENOENT') throw e; }
  }
  return removed;
}

/**
 * Every complete pair in a directory, newest first.
 *
 * Sorted by stem, which sorts chronologically because the stem is a fixed-width
 * UTC timestamp — no parsing, and no dependence on filesystem mtime, which a
 * copy or a restore from a host backup would rewrite.
 */
function listPairs(dir) {
  let names;
  try { names = fs.readdirSync(dir); }
  catch (e) { if (e.code === 'ENOENT') return []; throw e; }

  const stems = names
    .filter(n => n.endsWith('.rsc.gz'))
    .map(n => n.slice(0, -'.rsc.gz'.length))
    .filter(stem => names.includes(stem + '.backup'));

  return stems.sort().reverse().map(stem => {
    const stat = (p) => { try { return fs.statSync(p).size; } catch (_) { return 0; } };
    return {
      stem,
      rscBytes: stat(rscPath(dir, stem)),
      backupBytes: stat(backupPath(dir, stem)),
    };
  });
}

/** '2026-08-19T203521' back to epoch ms. Returns NaN for anything else. */
function _stemToMs(stem) {
  const m = /^(\d{4})-(\d{2})-(\d{2})T(\d{2})(\d{2})(\d{2})$/.exec(String(stem));
  if (!m) return NaN;
  return Date.UTC(+m[1], +m[2] - 1, +m[3], +m[4], +m[5], +m[6]);
}

/**
 * Which pairs retention should remove.
 *
 * Pure, so the rule can be tested without a filesystem — and so a preview can
 * never disagree with what the sweep actually deletes.
 *
 * Both limits apply and the stricter one wins, because they answer different
 * questions: keepCount bounds disk, keepDays bounds relevance. A zero or
 * missing limit means that limit is not applied at all.
 *
 * **The newest pair is never removed.** A router whose configuration has been
 * stable for longer than keepDays would otherwise age out its only restore
 * point precisely because nothing has gone wrong — the case where losing it
 * matters most. Pairs are written only when the configuration changed, so the
 * newest one is the current configuration however old it is.
 */
function selectForPruning(pairs, opts, now) {
  const keepCount = Number(opts && opts.keepCount) || 0;
  const keepDays = Number(opts && opts.keepDays) || 0;
  const sorted = pairs.slice().sort((a, b) => (a.stem < b.stem ? 1 : a.stem > b.stem ? -1 : 0));
  if (sorted.length <= 1) return [];

  const doomed = new Set();
  if (keepCount > 0) for (const p of sorted.slice(keepCount)) doomed.add(p.stem);
  if (keepDays > 0) {
    const cutoff = (now == null ? Date.now() : now) - keepDays * 86400000;
    for (const p of sorted) if (_stemToMs(p.stem) < cutoff) doomed.add(p.stem);
  }
  doomed.delete(sorted[0].stem);
  return sorted.filter(p => doomed.has(p.stem)).map(p => p.stem);
}

/** Total bytes a router's backups occupy, for the page's footer. */
function usageOf(dir) {
  return listPairs(dir).reduce((sum, p) => sum + p.rscBytes + p.backupBytes, 0);
}

module.exports = {
  baseDir,
  slugFor,
  dirFor,
  stemFor,
  rscPath,
  backupPath,
  ensureDir,
  writePair,
  readRsc,
  readBackup,
  hasPair,
  removePair,
  listPairs,
  selectForPruning,
  usageOf,
  _stemToMs,
};

File: internal/backups/store.go
package backups

// Where backups live on disk.
//
// ── A PAIR IS THE UNIT ──────────────────────────────────────────────────────
//
//	<data>/config-backups/<router-slug>/
//	    2026-08-19T203521.rsc.gz      gzipped export — diffable, no secrets
//	    2026-08-19T203521.backup      aes-sha256 binary — restorable
//
// The two share a stem and are always created, removed and counted together. A
// `.rsc` without its `.backup` is a diff you cannot act on; a `.backup` without
// its `.rsc` is a restore point you cannot inspect.
//
// ── THE DIRECTORY NAME IS A TRUST BOUNDARY ──────────────────────────────────
//
// It is derived from the router LABEL, which an operator types. `SlugFor` exists
// so that a label can never escape the base directory, and it is deliberately
// lossy rather than clever: lowercase, dashes, nothing else. Every traversal
// shape collapses — `../../etc/passwd` becomes `etc-passwd`, `..` becomes
// `router` — because the alternative is a sanitiser that has to be right about
// every encoding.
//
// The DATABASE does not trust the slug either: it records the directory that was
// used, and nothing resolves an old backup by re-deriving a slug from the current
// label. Renaming a router starts a new directory and leaves the old one
// findable.

import (
	"bytes"
	"compress/gzip"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

// unsafeInSlug is everything a directory name may not contain. Applied AFTER
// lower-casing, so an upper-case letter survives as its lower-case self.
var unsafeInSlug = regexp.MustCompile(`[^a-z0-9]+`)

var trimDashes = regexp.MustCompile(`^-+|-+$`)

// SlugFor is a directory name from a router label.
//
// A label of only punctuation slugs to nothing and falls back to a fixed name
// rather than producing "" and writing into the BASE directory itself.
//
// THE CUT TO 60 HAPPENS AFTER THE TRIM, NOT BEFORE, so a cut landing on a dash
// leaves a trailing dash and the original does not re-trim. Reproduced rather
// than tidied: a port that trimmed again would put the same router in a
// different directory from the one the database recorded.
func SlugFor(label string) string {
	s := strings.ToLower(label)
	s = unsafeInSlug.ReplaceAllString(s, "-")
	s = trimDashes.ReplaceAllString(s, "")
	// Bytes, not runes — but the string holds only [a-z0-9-] by now, so the two
	// are the same thing here and cannot cut a character in half.
	if len(s) > 60 {
		s = s[:60]
	}
	if s == "" {
		return "router"
	}
	return s
}

// BaseDir is resolved per call rather than frozen at package load.
//
// A constant meant the live test suite wrote real backup pairs into the real
// /data/config-backups and left a `test-router` directory on a production
// instance, which is how that was found. Taking the data directory as an
// argument lets a caller point it somewhere disposable.
func BaseDir(dataDir string) string {
	if dataDir == "" {
		dataDir = "/data"
	}
	return filepath.Join(dataDir, "config-backups")
}

// DirFor is a router's backup directory.
func DirFor(dataDir, slug string) string { return filepath.Join(BaseDir(dataDir), slug) }

// RscPath and BackupPath are the two halves of a pair.
func RscPath(dir, stem string) string    { return filepath.Join(dir, stem+".rsc.gz") }
func BackupPath(dir, stem string) string { return filepath.Join(dir, stem+".backup") }

// StemFor is the filename stem for a moment in time, IN UTC.
//
// UTC because a local-time stem repeats itself for an hour every autumn, and two
// backups that sort as equal are two backups that can overwrite each other.
// Seconds are included for the same reason at a smaller scale: a scheduled run
// and a manual one can land in the same minute.
//
// It must also produce a stem `stemToMs` can read back — retention parses these,
// and a stem this app writes but cannot re-read would never be aged out.
// `TestStemsRoundTrip` is what holds those two together.
func StemFor(ts int64) string {
	t := time.UnixMilli(ts).UTC()
	return t.Format("2006-01-02T150405")
}

// ── The pair on disk ────────────────────────────────────────────────────────

// WritePair writes both halves.
//
// THE BINARY LANDS FIRST, AND THE ORDER IS CRASH SAFETY RATHER THAN STYLE. If
// the process dies between the two writes, what survives is a `.backup` with no
// `.rsc` — and ListPairs ignores it, because a half-written pair is not a
// backup. The other order would leave a diffable export on disk claiming a
// restore point exists, and the failure would only surface when somebody tried
// to restore from it.
func WritePair(dir, stem, rscText string, binary []byte) (rscBytes, backupBytes int64, err error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return 0, 0, err
	}
	var buf bytes.Buffer
	zw := gzip.NewWriter(&buf)
	if _, err := zw.Write([]byte(rscText)); err != nil {
		return 0, 0, err
	}
	if err := zw.Close(); err != nil {
		return 0, 0, err
	}
	if err := os.WriteFile(BackupPath(dir, stem), binary, 0o600); err != nil {
		return 0, 0, err
	}
	if err := os.WriteFile(RscPath(dir, stem), buf.Bytes(), 0o600); err != nil {
		return 0, 0, err
	}
	return int64(buf.Len()), int64(len(binary)), nil
}

// ReadRsc is the stored export, gunzipped.
func ReadRsc(dir, stem string) (string, error) {
	f, err := os.Open(RscPath(dir, stem))
	if err != nil {
		return "", err
	}
	defer f.Close()
	zr, err := gzip.NewReader(f)
	if err != nil {
		return "", err
	}
	defer zr.Close()
	b, err := io.ReadAll(zr)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// ReadBackup is the stored binary, as it will be pushed back to the router.
//
// []byte, never a string routed through JSON — see rawbytes_test.go.
func ReadBackup(dir, stem string) ([]byte, error) {
	return os.ReadFile(BackupPath(dir, stem))
}

// HasPair reports whether BOTH halves are present. Neither is useful alone: a
// `.rsc` without its `.backup` is a diff you cannot act on, and a `.backup`
// without its `.rsc` is a restore point you cannot inspect.
func HasPair(dir, stem string) bool {
	if _, err := os.Stat(RscPath(dir, stem)); err != nil {
		return false
	}
	_, err := os.Stat(BackupPath(dir, stem))
	return err == nil
}

// RemovePair removes a pair and reports how many files went.
//
// MISSING HALVES ARE NOT AN ERROR. Pruning has to be idempotent, because it runs
// after a crash as readily as after a success — and a crash is exactly what
// leaves one half behind.
func RemovePair(dir, stem string) (int, error) {
	removed := 0
	for _, p := range []string{RscPath(dir, stem), BackupPath(dir, stem)} {
		err := os.Remove(p)
		switch {
		case err == nil:
			removed++
		case os.IsNotExist(err):
			// Fine.
		default:
			return removed, err
		}
	}
	return removed, nil
}

// ListPairs is every COMPLETE pair in a directory, newest first.
//
// Sorted by stem, which sorts chronologically because the stem is a fixed-width
// UTC timestamp — no parsing, and no dependence on filesystem mtime, which a
// copy or a restore from a host backup would rewrite.
//
// A missing directory is an empty list rather than an error: a router that has
// never been backed up has no directory, and that is not a fault.
func ListPairs(dir string) ([]Pair, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if os.IsNotExist(err) {
			return []Pair{}, nil
		}
		return nil, err
	}
	names := map[string]bool{}
	for _, e := range entries {
		names[e.Name()] = true
	}

	out := []Pair{}
	for name := range names {
		stem, ok := strings.CutSuffix(name, ".rsc.gz")
		if !ok || !names[stem+".backup"] {
			continue
		}
		p := Pair{Stem: stem}
		if fi, err := os.Stat(RscPath(dir, stem)); err == nil {
			p.RscBytes = fi.Size()
		}
		if fi, err := os.Stat(BackupPath(dir, stem)); err == nil {
			p.BackupBytes = fi.Size()
		}
		out = append(out, p)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Stem > out[j].Stem })
	return out, nil
}

// UsageOf is the total bytes a router's backups occupy.
func UsageOf(dir string) (int64, error) {
	pairs, err := ListPairs(dir)
	if err != nil {
		return 0, err
	}
	var sum int64
	for _, p := range pairs {
		sum += p.RscBytes + p.BackupBytes
	}
	return sum, nil
}
